Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/Exporters/TemplatedExporter/TemplatedExporterPart.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#if UNITY_EDITOR
+using Microsoft.Build.Unity.ProjectGeneration.Templates;
+using System;
+
+namespace Microsoft.Build.Unity.ProjectGeneration.Exporters.TemplatedExporter
+{
+    /// <summary>
+    /// Base class for template based exporters.
+    /// </summary>
+    internal class TemplatedExporterPart
+    {
+        private readonly ITemplatePart template;
+        protected readonly TemplateReplacementSet replacementSet;
+
+        protected TemplatedExporterPart(ITemplatePart template, TemplateReplacementSet replacementSet)
+        {
+            this.template = template;
+            this.replacementSet = replacementSet;
+        }
+
+        /// <summary>
+        /// Updates the value of the field and the token (if the field has changed).
+        /// </summary>
+        /// <param name="field">The ref field to update.</param>
+        /// <param name="value">The value to update to.</param>
+        /// <param name="token">The token key.</param>
+        protected void UpdateToken(ref string field, string value, string token)
+        {
+            UpdateToken(ref field, value, token, t => t);
+        }
+
+        /// <summary>
+        /// Updates the value of the field and the token (if the field has changed).
+        /// </summary>
+        /// <typeparam name="T">The generic type of the field and value.</typeparam>
+        /// <param name="field">The ref field to update.</param>
+        /// <param name="value">The value to update to.</param>
+        /// <param name="token">The token key.</param>
+        /// <param name="toStringFunc">The conversion to string value function.</param>
+        protected void UpdateToken<T>(ref T field, T value, string token, Func<T, string> toStringFunc)
+        {
+            if (!Equals(field, value))
+            {
+                field = value;
+                template.Tokens[token].AssignValue(replacementSet, toStringFunc(value));
+            }
```

Comment:
we should asses whether the token exists in this Tokens collection before calling AssignValue

Reply:
Possibly, but since this exporter is tied to the templates, we assume otherwise it's a bug.

Reply:
it might be easier to debug a logged error with the token that was missing compared to an exception

---
In reply to: [368175853](https://github.com/microsoft/MSBuildForUnity/pull/109#discussion_r368175853) [](ancestors = 368175853)